### DISCUSSION TITLE
Bulk Import: how to benifit from machine learning?

### DISCUSSION THREAD

**Discussion by hwine** [paperless-ngx/paperless-ngx#8866] [Support] [answered]
Noob to Paperless, but having it up and running was simple thanks to great doc* (docs & docker).

Is there a recommended workflow for importing "folder sorted" existing digital material? I'm looking for one that enables the paperless machine learning as soon as possible.

What I've tried is:
- drop a folder into the 'consume' directory, which imports the docs and applies the `Inbox` tag
  - (well I wrote a script to move items out of the folders, but that's another question)
- manually process the Inbox to add the tags and correspondent I want, which are largely the same for all docs from the folder tree

What I was expecting was for the machine learning to "kick in" at some point, and offer better choices.
- that doesn't appear to happen

I can think of lots of reasons why it isn't working the way I expect (I'm somewhat an expert on PEBKAC), so I'm hoping y'all can help me improve my experience. After reading the docs, I am not clear on _when_ the machine learning both "learns" and "applies learning". 

The above is all somewhat abstract, so here are some more detailed questions that I think will help me form a useful mental model of paperless' machine learning.

1. Will paperless ever apply something-newly-learned to existing documents-in-inbox?
2. Does paperless perform "training runs" on a scheduled basis? I.e. would I see a benefit from categorizing a few documents, waiting some time period, then consuming more?
3. The cautions on using the "reprocess document" action sound rather severe. Is there any significant impact to using that action on "untouched" items in the inbox?
4. What is a reasonable expectation on when I should see the effects of training? ("it depends" is a fine answer)

Related, is there a way to set up a "manually triggered workflow" to do the 3 or 4 tagging steps when I'm working on a "folder" of similar documents?  E.g. a (temporary) custom action/workflow to assign a correspondent & a couple of tags.

TIA.

**Comment by shamoon** [comment]
Theres a lot of questions here, but take a look at https://docs.paperless-ngx.com/advanced_usage/#matching / https://docs.paperless-ngx.com/advanced_usage/#automatic-matching

Automatic matching needs a set of documents that are correctly tagged to learn from, and the classifier needs to be trained on these (happens periodically automatically). So basically you will need to initially start with manually tagging a set and then can try auto after that. Again, just like most ML algorithms, the larger the [correct] training data set the better the results, roughly speaking.

**Comment by hwine** [comment]
Shamoon -- thanks, but I'd already read [that page](https://docs.paperless-ngx.com/advanced_usage/#matching), and still had all the questions.

I'll assume that the feature won't kick in until there's a lot more documents (I only have ~100 so far).

I did discover that one can "select similar" then edit-as-a-batch, which does reduce the initial workload some.

**Comment by ghost** [comment]
What i did in the paßt is to Import online 10, set the Metadata, and Importen another ten to see if it hast already learned so far.

**Comment by luziferius** [comment, marked as answer]
To get the initial training data going, the system needs a verity of different documents. And then do ML training on that. 

1. I don't think so. But there's a re-process button in the actions menu. So select all documents in the inbox, and let paperless re-process them.
2. Once per hour. That's what I see in the log file. Edit: Or manually via `document_create_classifier` as stated in [https://docs.paperless-ngx.com/administration/#managing-the-automatic-matching-algorithm](https://docs.paperless-ngx.com/administration/#managing-the-automatic-matching-algorithm)
3. No idea on that one. But since there is no manual editing done yet, I think at worst you waste CPU cycles by redoing OCR and stuff.
4. When I started, the first hundred documents or so had many errors, but that was on a year old version. I'm now at around 500 documents, and it's gotten way better. (Also new features, like suggesting alternative dates found in the text helps a lot with the document date field)

The learning needs both positive and negative examples, otherwise the false positive or false negative rate will shoot through the roof. An initial import of pre-sorted documents does not provide that, as it is biased towards the sorting criterion used by the directory structure, so expect bad behavior initially.

Paperless rejects duplicates, so you can take advantage of that:
Select a random set of documents from your existing archive and add them to paperless. Then tag them correctly and remove the Inbox tag. Once the inbox tag is cleared, it is available as ML training data. Manually re-trigger the classifier training or wait an hour, as that's the default training interval. Then do several more rounds of training like this.
Once the hit rate is sufficiently high, dump everything in. Paperless will complain about the duplicates and ignore them, so you can discard these duplicate file errors.

**Reply by stumpylog** [reply]
https://docs.paperless-ngx.com/administration/#managing-the-automatic-matching-algorithm

**Reply by luziferius** [reply]
Thanks. Amended my comment with the link.

**Comment by github-actions** [comment]
This discussion has been automatically closed because it was marked as answered. Please see our [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/dev/CONTRIBUTING.md#automatic-repository-maintenance) for more details.

**Comment by github-actions** [comment]
This discussion has been automatically locked since there has not been any recent activity after it was closed. Please open a new discussion for related concerns. See our [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/dev/CONTRIBUTING.md#automatic-repository-maintenance) for more details.
